cloud/aws/services/ec2/natgateway.go

```diff
@@ -0,0 +1,126 @@
+package ec2
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/cluster-api-provider-aws/cloud/aws/providerconfig/v1alpha1"
+)
+
+func (s *Service) reconcileNatGateways(subnets []*v1alpha1.Subnet, vpc *v1alpha1.VPC) error {
+	if !s.containsPrivateSubnets(subnets) {
+		return nil
+	}
+
+	existing, err := s.describeNatGateways(vpc.ID)
+	if err != nil {
+		return err
+	}
+
+	for _, sn := range subnets {
+		if sn.ID == "" || !sn.IsPublic {
+			continue
+		}
+		if _, ok := existing[sn.ID]; ok {
+			continue
+		}
+
+		_, err := s.createNatGateway(sn.ID)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Service) describeNatGateways(vpcID string) (map[string]*ec2.NatGateway, error) {
```

Comment:
Little nitpick around naming: describeNATGatewaysBySubnet is more explicit for users. For example, I had to read through the code to understand that the key of the map was the subnetId.

Reply:
Have renamed. Thanks.